array_api_compat/torch/_aliases.py

```diff
@@ -135,8 +135,12 @@ def result_type(*arrays_and_dtypes: Union[array, Dtype, bool, int, float, comple
         return result_type(arrays_and_dtypes[0], result_type(*arrays_and_dtypes[1:]))
 
     x, y = arrays_and_dtypes
-    if isinstance(x, _py_scalars) or isinstance(y, _py_scalars):
-        return torch.result_type(x, y)
+    if isinstance(x, _py_scalars):
+        if isinstance(y, _py_scalars):
+            raise ValueError("At least one array or dtype is required.")
```

Comment:
If I'm understanding the recursion correctly, won't this fail if, for instance, `xp.result_type(xp.float64, 1, 2)` because neither of the last two arguments is an array or dtype?

Reply:
True, but you must fundamentally change the logic here anyway.  While torch avoids tricky promotions (unlike NumPy) and promotions are associative because of that, you are currently breaking that by adding scalar promotion rules.
That isn't hard to fix, because you can fix it by fixing the binary promotion to:
```
promote(pythong_float, torch_integer) -> python_float
```

I.e. the recursive part of the algorithm *must* drag around `python_float` as a proper dtype.  Only at the very end can you convert it to the default float.

Reply:
Given that one way or other, we need to add an arbitrary rule, and we cannot fully satisfy all of Array API promotion rules, pytorch promotion rules and associativity, how about always sorting the inputs to put all scalars to the left. This way, the pytorch addition only kicks up in a binary promotion ("a conforming library may support additional type promotion rules", so pytorch does---for binary promotions only), and we definitely have associativity. Thoughts?

Reply:
I don't understand the first premise of needing an arbitrary rule or that you can't satisfy everything...

... but it doesn't matter?  Currently, torch (by its rules) _is_ associative, I believe.  The only odd one out is the custom behavior for `float` due to `int_dtype + float -> default_float` being wrong (breaking associativity).

And sure, since that is all you need to deal with, you can deal with it by ensuring Python scalar promotions happen last.